Deserialize dictionary key with TypeDescriptor

Joy-less opened the pull request:
`BsonDocument` only supports `string` keys. Since `Dictionary` is serialized as `BsonDocument`, certain types don't work as keys.

This pull request converts the keys to the correct type when deserializing, using the method in #588.

For example, the key `d396d0e8-2763-4ad6-a1c0-f19f254576ec` can now be deserialized as a `Guid`.

This isn't really ideal but it does the job.

Fixes #546 and #2161. Replaces #588.

Joy-less committed:
Deserialize dictionary key with TypeDescriptor

JKamsker commented:
Hi @Joy-less 

Great job on the changes! Could you also implement and ensure the following works as expected?

```csharp
public class DataContainer
{
    public Dictionary<(int, string), string> DataDict { get; set; }
}
```

Additionally, as mentioned earlier on Discord, reviewing a PR becomes challenging when it includes unrelated changes (e.g., refactor-renames). For future submissions, keeping PRs focused on a single purpose would help simplify the review process.

Thank you for your efforts on this!

Joy-less commented:
> Would it be possible to also implement and ensure the following works as expected?
> 
> ```cs
> public class DataContainer
> {
>     public Dictionary<(int, string), string> DataDict { get; set; }
> }
> ```

This does not work as expected. The key should be serialized as JSON for something complex like this. However, that would break compatibility, so it might be a good idea to assume it's a string key if it fails to deserialize as JSON.

Joy-less committed:
Serialize/deserialize dictionary keys as JSON

Joy-less committed:
Avoid serializing string keys as JSON

Joy-less committed:
Don't truncate DateTime

Joy-less committed:
Add comments and improve formatting

Joy-less committed:
Undo "don't truncate DateTime".

Otherwise the BufferWriter is not the same as the JsonSerializer.

Joy-less commented:
I've changed the pull request to serialize the key as JSON if it's not a string.
One of the tests was failing because of how LiteDB truncates DateTimes to the nearest millisecond. To keep `JsonSerializer` consistent with `BufferWriter`, I changed the test so now it accepts the truncated value as well.
All tests passed now.

JKamsker commented:
@Joy-less 
Thanks for the changes! 
Can you think of an scenario where the deserialization doesnt work on databases created with a prior (released) version of litedb?

Joy-less committed:
Add some tests to ensure enums are deserialized correctly.

JKamsker commented on LiteDB/Client/Mapper/BsonMapper.Serialize.cs:
```diff
@@ -128,52 +126,55 @@ internal BsonValue Serialize(Type type, object obj, int depth)
-                var skey = key.ToString();
+                string stringKey = entry.Key as string
+                    // Serialize key as JSON to support any key type
+                    ?? JsonSerializer.Serialize(Serialize(entry.Key));
```
```csharp
TupleDict = new()
  {
      [(2, "xxx")] = "test",
  },
  ```
from the tests produces
```csharp
System.ValueTuple`2[[System.Int32, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib
```

JKamsker replied:
```csharp
TupleDict = new()
  {
      [(2, "xxx")] = "test",
  },
  ```
from the tests produces
```csharp
System.ValueTuple`2[[System.Int32, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib
```
